src/components/control-component.js

```diff
@@ -89,564 +89,578 @@ const propTypes = {
   dynamic: PropTypes.bool,
 };
 
-function mapStateToProps(state, props) {
-  const {
-    model,
-    getter = _get,
-    controlProps = omit(props, Object.keys(propTypes)),
-  } = props;
-
-  const modelString = getModel(model, state);
-  const fieldValue = getFieldFromState(state, modelString)
-    || initialFieldState;
-
-  return {
-    model: modelString,
-    modelValue: getter(state, modelString),
-    fieldValue,
-    controlProps,
-  };
-}
-
-function isReadOnlyValue(controlProps) {
-  return ~['radio', 'checkbox'].indexOf(controlProps.type);
-}
+function createControlClass(customControlPropsMap = {}, defaultProps = {}) {
 
-const emptyControlProps = {};
-
-class Control extends Component {
-  constructor(props) {
-    super(props);
+  const controlPropsMap = {
+    ...defaultControlPropsMap,
+    ...customControlPropsMap,
+  };
 
-    this.getChangeAction = this.getChangeAction.bind(this);
-    this.getValidateAction = this.getValidateAction.bind(this);
+  function mapStateToProps(state, props) {
+    const {
+      model,
+      getter = Control.defaultProps.getter,
```

Comment:
This was the most crucial change in making this component compatible with immutable